benchmark/cublas/blas1/axpy.cpp

```diff
@@ -0,0 +1,159 @@
+/***************************************************************************
+ *
+ *  @license
+ *  Copyright (C) Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename axpy.cpp
+ *
+ **************************************************************************/
+
+#include "../utils.hpp"
+
+template <typename scalar_t>
+std::string get_name(int size) {
+  std::ostringstream str{};
+  str << "BM_Axpy<" << blas_benchmark::utils::get_type_name<scalar_t>() << ">/";
+  str << size;
+  return str.str();
+}
+
+template <typename scalar_t, typename... args_t>
+static inline void cublas_routine(args_t&&... args) {
+  if constexpr (std::is_same_v<scalar_t, float>)
+    CUBLAS_CHECK(cublasSaxpy(std::forward<args_t>(args)...));
+  else if constexpr (std::is_same_v<scalar_t, double>)
+    CUBLAS_CHECK(cublasDaxpy(std::forward(args)...));
+  return;
+}
+
+template <typename scalar_t>
+void run(benchmark::State& state, cublasHandle_t* cuda_handle_ptr, index_t size,
+         bool* success) {
+  // Google-benchmark counters are double.
+  double size_d = static_cast<double>(size);
+  state.counters["size"] = size_d;
+  state.counters["n_fl_ops"] = 2.0 * size_d;
+  state.counters["bytes_processed"] = 3.0 * size_d * sizeof(scalar_t);
```

Comment:
I can see that these numbers are setting probably for all AXPY benchmarks. Similarly, for other BLAS routine benchmarks these numbers are the same. Can we put per routine config in the benchmark utils so that we can call them and don't repeat them every time we are writing the benchmark? In that case, it would be less error prune as well.

Reply:
If these counters need to record the number of floating-point operations and memory transactions performed by a routine, they cannot be the same for all BLAS1 routines: e.g., for AXPY `"bytes_processed"` is equal to 3 times the array size, for IAMAX is 1 time the array size (+1 int, if we want to consider the output).